### DISCUSSION TITLE
Struggling to get twitter auth setup

### DISCUSSION THREAD

**Discussion by secondubly** [umputun/remark42#1280] [Q&A]
As the title says, I'm struggling to get my twitter auth setup for commenting on my blog. I've followed the instructions on the docs as close as possible but I think things have changed since it was written and I don't know what I'm missing. I have my consumer api key and consumer secret key - but when I log in with my twitter account I get this in the debug logs:

```
remark42  | 2022/02/20 23:26:54.011 [DEBUG] {provider/oauth1.go:33 provider.Oauth1Handler.LoginHandler} login with twitter
remark42  | 2022/02/20 23:26:54.216 [INFO]  {logger/logger.go:130 logger.(*Middleware).Handler.func1.1} GET - /auth/twitter/login?from=https://[redacted]/web/iframe.html?selfClose&site=remark - [redacted] - b3d1f97212e7 - 302 (102) - 204.604986ms
remark42  | 2022/02/20 23:26:57.933 [DEBUG] {provider/oauth1.go:127 provider.Oauth1Handler.AuthHandler} got raw user info map[errors:[map[code:453 message:You currently have Essential access which includes access to Twitter API v2 endpoints only. If you need access to this endpoint, you’ll need to apply for Elevated access via the Developer Portal. You can learn more here: https://developer.twitter.com/en/docs/twitter-api/getting-started/about-twitter-api#v2-access-leve]]]
remark42  | 2022/02/20 23:26:57.947 [DEBUG] {avatar/avatar.go:157 avatar.(*Proxy).resize} avatar resize(): limit should be greater than 0
remark42  | 2022/02/20 23:26:57.948 [DEBUG] {avatar/avatar.go:51 avatar.(*Proxy).Put.func1} saved identicon avatar to 77c542aeab0849ee4775b3aa97b691be762585d8.image, user ""
remark42  | 2022/02/20 23:26:57.948 [WARN]  {cmd/server.go:1076 cmd.(*ServerCommand).getAuthenticator.func2} can't read email for twitter_da39a3ee5e6b4b0d3255bfef95601890afd80709, site "remark" not found
remark42  | 2022/02/20 23:26:57.949 [DEBUG] {provider/oauth1.go:156 provider.Oauth1Handler.AuthHandler} user info {Name: ID:twitter_da39a3ee5e6b4b0d3255bfef95601890afd80709 Picture:https://[redacted]/api/v1/avatar/77c542aeab0849ee4775b3aa97b691be762585d8.image Audience: IP: Email: Attributes:map[admin:true blocked:false] Role:}
remark42  | 2022/02/20 23:26:57.949 [INFO]  {logger/logger.go:130 logger.(*Middleware).Handler.func1.1} GET - /auth/twitter/callback?oauth_token=h3WGMwAAAAABZd2CAAABfxq_OaQ&oauth_verifier=UpLc8pvl65pB0Qgt0QReE5P7XT9DP6hK - [redacted] - b3d1f97212e7 - 307 (92) - 533.946572ms
remark42  | 2022/02/20 23:26:58.087 [WARN]  {api/rest_private.go:249 api.(*private).userInfoCtrl} can't read email for twitter_da39a3ee5e6b4b0d3255bfef95601890afd80709, site "remark" not found
remark42  | 2022/02/20 23:26:58.088 [INFO]  {logger/logger.go:130 logger.(*Middleware).Handler.func1.1} GET - /api/v1/user?site=remark - [redacted] - b3d1f97212e7 - 200 (204) - 898.004µs
```

I hyper-focused in on this line: `got raw user info map[errors:[map[code:453 message:You currently have Essential access which includes access to Twitter API v2 endpoints only. If you need access to this endpoint, you’ll need to apply for Elevated access via the Developer Portal. You can learn more here: https://developer.twitter.com/en/docs/twitter-api/getting-started/about-twitter-api#v2-access-leve]]]` but applying for Elevated access seems unnecessary, so am I missing something? I've tried both 1.0 oauth and 2.0 oauth to no success either way. I'd also be more than happy to update the docs if it would help make things easier to understand for those coming after me.

**Comment by umputun** [comment]
I'm not sure about "Elevated access" as at the time we have added twitter support oauth1 was the only way with twitter and this is what [go-pkgz/auth](https://github.com/go-pkgz/auth) implements. See [this pr](https://github.com/go-pkgz/auth/pull/33) for details.

Btw, none of my twitter "apps" require elevated access, but maybe this is a thing for new apps, not sure.

The best way to proceed here is to open a ticket in go-pkgz/auth and mention the author of that PR, so he could check it and hopefully address the issue. If Twitter is really supporting oauth2 these days, adding it to auth providers should be a relatively easy task.

**Comment by secondubly** [comment]
Did a bit more research last night/this morning and found this regarding the Elevated/Essential Access confusion (seen here https://developer.twitter.com/en/docs/twitter-api/getting-started/about-twitter-api):

> If you were approved for a developer account before November 15th, 2021, you were automatically converted to [Elevated access](https://developer.twitter.com/en/docs/twitter-api/getting-started/about-twitter-api#v2-access-level). This means that your existing [Apps](https://developer.twitter.com/en/docs/apps) can continue to be used to make requests to standard v1.1, premium v1.1, and enterprise endpoints, and that all of your user Access Tokens are still valid.
> 
> If you would like to start using v2 endpoints, you will need to attach an App to a [Project](https://developer.twitter.com/en/docs/projects) and use the credentials from that App when making requests to v2 endpoints.

So for those who setup their accounts before that date, their apps will work fine - anyone coming to this project after that and wanting to use twitter auth with be required to use oauth 2 unless the apply for elevated access (which puts you on a waitlist) - I'll go ahead and move over to go-pkgz/auth and continue the conversation there, just wanted to leave this here in case anyone stops by and is confused on all the run-around.

**Reply by paskal** [reply]
If someone will figure out how to properly request currently working Twitter auth, please update https://github.com/go-pkgz/auth#twitter-auth-provider and https://github.com/umputun/remark42/blob/master/site/src/docs/configuration/authorization/index.md#twitter appropriately to simplify the life of other users. Thanks!
